src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -79,6 +80,8 @@ public virtual IDisposable Start()
                         contextAccessor.HttpContext = httpContext;
                         await application(httpContext);
                     }
+
+                    return httpContext;
```

Comment:
Is this so httpContext gets disposed? What are we waiting for?

I would just wrap the call to `await application(httpContext)` in a using statement. Once the Task returned from the application is completed, you should be safe to dispose the httpContext.